### ISSUE TITLE
GHOST estimator should return _children_ of the heaviest subtree.

### ISSUE THREAD

**Issue by afck** [cbc-casper/cbc-casper-paper#28]
Definition 4.31 (Casper the Friendly Ghost) gives the estimator as:
ℰ(σ) = GHOST({g}, σ)
which is the set of all blocks occurring in σ that are reachable from the genesis block g by always picking a child with maximum score. In particular, all elements of ℰ(σ) are blocks that are already the estimate of a message in σ, so that estimator wouldn't allow creating new blocks.

Should that definition be changed to instead allow all conceivable _children_ of the elements of GHOST({g}, σ)? Because the next estimate should not be a block that has been proposed before, but a new child.

**Comment by adiasg** [comment]
The purpose of the estimator here is to return the tip blocks winning under the GHOST rule. It doesn't return new blocks for proposal from a given protocol state.

We should introduce a dichotomy in the usage of estimator: generate new blocks vs. verify blocks in protocol state. The estimator function that generates new blocks will be useful in describing validator behavior, and hence isn't mentioned in this draft (which concerns itself with reasoning about things after we reach a particular protocol state).

**Comment by afck** [comment]
I understand that the GHOST function is meant to return the winning tips.
But the _estimator_ ℰ(σ) defines what consitutes a legal message: A message's estimate is required to be an element of ℰ(σ), where σ is the message's justification (Definition 2.7). That means if in σ, all estimates are g, then with the above estimator, ℰ(σ) = {g}, and therefore the new message's estimate also must be g. Thus _all_ messages in _all_ protocol states would have estimate g.

If I'm not misreading this, I think the definition 4.31 should be something like this instead:
ℰ(σ) = {b | Prev(b) ∈ GHOST({g}, σ)}

**Comment by HarikrishnanBalagopal** [comment]
> 
> 
> I understand that the GHOST function is meant to return the winning tips.
> But the _estimator_ ℰ(σ) defines what consitutes a legal message: A message's estimate is required to be an element of ℰ(σ), where σ is the message's justification (Definition 2.7). That means if in σ, all estimates are g, then with the above estimator, ℰ(σ) = {g}, and therefore the new message's estimate also must be g. Thus _all_ messages in _all_ protocol states would have estimate g.
> 
> If I'm not misreading this, I think the definition 4.31 should be something like this instead:
> ℰ(σ) = {b | Prev(b) ∈ GHOST({g}, σ)}

Yes but it's actually even worse than that because the way it is defined now, it goes into an infinite recursion:
![image](https://user-images.githubusercontent.com/20921177/110242719-52b27a00-7f7d-11eb-8048-6b42bf017d4a.png)
![image](https://user-images.githubusercontent.com/20921177/110242701-42020400-7f7d-11eb-8c53-fce8fd3f5799.png)
![image](https://user-images.githubusercontent.com/20921177/110242112-84761180-7f7a-11eb-9706-6d829fb94b52.png)

1. We call `GHOST({g}, \sigma)` where `\sigma = {(g, v1, {}), (g, v2, {}), (g, v3, {})}`
2. The children of `g` is `{g}` (because prev(g) = g)
3. So best children will also be `{g}`.
4. So you end up calling `GHOST({g}, \sigma)` again.

Sidenote:
@afck @adiasg  are you guys maintainers on this repo? There are some open issues and pull requests that are from 2018 and 2019 that haven't been resolved yet. Has the repo been abandonded?

**Comment by afck** [comment]
No, I'm not a maintainer.

**Comment by adiasg** [comment]
I'm not a current maintainer. Unaware of the status of this repo.
